client/session.go

```diff
@@ -1912,17 +1994,135 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
 				Tail: segments.Unmerged[i].Tail,
 			})
 		}
-		encoder, err := r.mergeReaders(start, readers)
+		encoder, err := b.mergeReaders(start, readers)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
 		// Set the block data
 		result.Reset(start, encoder.Discard())
 
 	default:
-		return errSessionBadBlockResultFromPeer
+		return nil, errSessionBadBlockResultFromPeer
+	}
+
+	return result, nil
+}
+
+type streamBlocksResult struct {
+	baseBlocksResult
+	outputCh chan<- peerBlocksDatapoint
+}
+
+func newStreamBlocksResult(
+	opts Options,
+	resultOpts result.Options,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	outputCh chan<- peerBlocksDatapoint,
+) *streamBlocksResult {
+	blockOpts := resultOpts.DatabaseBlockOptions()
+	return &streamBlocksResult{
+		baseBlocksResult: baseBlocksResult{
+			blockOpts:               blockOpts,
+			blockAllocSize:          blockOpts.DatabaseBlockAllocSize(),
+			contextPool:             opts.ContextPool(),
+			encoderPool:             blockOpts.EncoderPool(),
+			multiReaderIteratorPool: multiReaderIteratorPool,
+		},
+		outputCh: outputCh,
+	}
+}
+
+type peerBlocksDatapoint struct {
+	id    ts.ID
+	peer  topology.Host
+	block block.DatabaseBlock
+}
+
+func (s *streamBlocksResult) addBlockFromPeer(id ts.ID, peer topology.Host, block *rpc.Block) error {
+	result, err := s.getDatabaseBlock(block)
+	if err != nil {
+		return err
+	}
+	s.outputCh <- peerBlocksDatapoint{
+		id:    id,
+		peer:  peer,
+		block: result,
+	}
+	return nil
+}
+
+type peerBlocksIter struct {
+	inputCh <-chan peerBlocksDatapoint
+	errCh   <-chan error
+	current peerBlocksDatapoint
+	err     error
+	done    bool
+}
+
+func newPeerBlocksIter(
+	inputC <-chan peerBlocksDatapoint,
+	errC <-chan error,
+) *peerBlocksIter {
+	return &peerBlocksIter{
+		inputCh: inputC,
+		errCh:   errC,
+	}
+}
+
+func (it *peerBlocksIter) Current() (topology.Host, ts.ID, block.DatabaseBlock) {
+	return it.current.peer, it.current.id, it.current.block
+}
+
+func (it *peerBlocksIter) Err() error {
+	return it.err
+}
+
+func (it *peerBlocksIter) Next() bool {
+	if it.done || it.err != nil {
+		return false
+	}
+	m, more := <-it.inputCh
 
+	if !more {
+		it.err = <-it.errCh
+		it.done = true
+		return false
+	}
+
+	it.current = m
+	return true
+}
+
+type bulkBlocksResult struct {
+	sync.RWMutex
+	baseBlocksResult
+	result result.ShardResult
+}
+
+func newBulkBlocksResult(
+	opts Options,
+	resultOpts result.Options,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+) *bulkBlocksResult {
+	blockOpts := resultOpts.DatabaseBlockOptions()
+	return &bulkBlocksResult{
+		baseBlocksResult: baseBlocksResult{
```

Comment:
We can extract this constructing of `baseBlocksResult` into a `newBaseBlocksResult(...)` you can call from both `streamBlocksResult` and `bulkBlocksResult` constructors perhaps?